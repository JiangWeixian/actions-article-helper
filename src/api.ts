import { ChatGPTAPI } from 'chatgpt'
import { fetch } from 'cross-fetch'

globalThis.fetch = fetch

export const createChatGPTAPI = (apiKey: string) => {
  return new ChatGPTAPI({
    apiKey,
  })
}
