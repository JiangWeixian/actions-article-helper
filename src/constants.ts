import { template } from 'lodash-es'

export const DEBUG_KEY = 'neo:article-helper'
export const COMMENT_AUTHOR = new Set(['github-actions[bot]'])
export const prefix = '<!--article-helper-->'
export const prompts = {
  // refs: https://github.com/f/awesome-chatgpt-prompts#act-as-an-english-translator-and-improver
  check: template('I want you to act as an English translator, spelling corrector and improver. I will speak to you in any language and you will detect the language, translate it and answer in the corrected and improved version of my text, in English. I want you to replace my simplified A0-level words and sentences with more beautiful and elegant, upper level English words and sentences. Keep the meaning same, but make them more literary. I want you to only reply the correction, the improvements and nothing else, do not write explanations. My markdown format article is "<%= content %>"'),
  // refs: https://wfhbrian.com/the-best-way-to-summarize-a-paragraph-using-gpt-3/
  summarize: template('Write short introduction with high source compression, removes stop words and summarizes the article whilst retaining meaning. The result is the shortest possible summary that retains all of the original meaning and context of the article.'),
}
export const codeBlockRE = /`{3}[\s\S]*?`{3}/g
