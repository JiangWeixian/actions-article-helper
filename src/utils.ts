import { get_encoding } from '@dqbd/tiktoken'
import { title } from 'functional-md'

import { codeBlockRE, prefix } from './constants'

const tokenizer = get_encoding('cl100k_base')

function encode(input: string) {
  return tokenizer.encode(input)
}

// refs: https://github.com/transitive-bullshit/chatgpt-api/blob/0ed1ad003568eba776da2dbd78a4edc73f8acad3/src/chatgpt-api.ts#L410
export const getTokenCount = async (text: string) => {
  text = text.replace(/<\|endoftext\|>/g, '')
  return encode(text).length
}

/**
 * @description Preprocess markdown content, e.g.
 * - remove code blocks
 */
export const preprocess = (content: string, { removeCodeblocks = false }: { removeCodeblocks: boolean } = { removeCodeblocks: false }) => {
  let resolvedContent = content
  if (removeCodeblocks) {
    resolvedContent = resolvedContent.replace(codeBlockRE, '```\n```')
  }
  return resolvedContent
}

const withLeadPrefix = (body: string) => {
  return `${prefix}\n${new Date()}\n${body}`
}

export const formatComment = ({ article, summary }: { article: string; summary: string }) => {
  const comment = `
  ${title({ heading: 1, children: 'Summary' })}
  ${summary}
  ${title({ heading: 1, children: 'Article' })}
  ${article}
  `
  return withLeadPrefix(comment)
}
