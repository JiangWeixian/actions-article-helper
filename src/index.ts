import * as core from '@actions/core'
import * as github from '@actions/github'
import Debug from 'debug'
import matter from 'gray-matter'

import { createChatGPTAPI } from './api'
import { COMMENT_AUTHOR, DEBUG_KEY, prefix, prompts } from './constants'
import { formatComment } from './utils'

const debug = Debug(DEBUG_KEY)

export interface Comment {
  id: number
  body?: string
  user: {
    login: string
  } | null
  created_at: string
}

const findComment = (comments: Comment[]) => {
  return comments.find(comment => comment.user?.login && COMMENT_AUTHOR.has(comment.user?.login) && comment.body?.includes(prefix))
}

const parseArticle = (body: string) => {
  const meta = matter(body, { delimiters: core.getInput('delimiters') ? JSON.parse(core.getInput('delimiters')) : undefined })
  return {
    ...meta,
    data: meta.data,
  }
}

async function main() {
  try {
    debug('start actions')
    const token = process.env.GITHUB_TOKEN
    const apiKey = process.env.OPENAI_API_KEY
    if (!token || !apiKey) {
      throw new Error('Github token and open api key <https://platform.openai.com/account/api-keys> is required.')
    }
    const octokit = github.getOctokit(token)
    const ownerOnly = core.getInput('ownerOnly') ?? true
    debug('Option: ownerOnly %s', ownerOnly)
    const { owner, repo } = github.context.repo
    const { number, owner: issueOwner } = github.context.issue
    if (issueOwner !== owner && ownerOnly) {
      throw new Error('Only owner can trigger this action.')
    }
    debug('issue context %O', github.context.issue)
    const issue = await octokit.rest.issues.get({
      issue_number: number,
      owner,
      repo,
    })
    if (!issue.data.body) {
      core.info('issue body is empty, skip.')
      return
    }
    const issueBody = parseArticle(issue.data.body!)
    const articlePrompt = prompts.check({ content: issueBody.content })
    const chatgptApi = await createChatGPTAPI(apiKey, { article: articlePrompt })
    debug('start check grammar...')
    const articleResult = await chatgptApi.sendMessage(articlePrompt, {
      stream: true,
    })
    debug('chatgpt response for grammar checker %o', articleResult)
    debug('start summarize article...')
    const summaryPrompt = prompts.summarize()
    const summaryResult = await chatgptApi.sendMessage(summaryPrompt, {
      stream: true,
      parentMessageId: articleResult.id,
    })
    debug('chatgpt response for summary %o', summaryResult)
    // list all <number> comments to find specific comments
    const comments = await octokit.rest.issues.listComments({
      owner,
      repo,
      issue_number: number,
    })
    debug('all comments %o from issue %s', comments.data, number)
    // find comments by bot<github_actioins> start with `<!--article-helper-->`
    const comment = findComment(comments.data)
    debug('find comment %o', comment)
    if (comment) {
      await octokit.rest.issues.updateComment({
        owner,
        repo,
        comment_id: comment?.id,
        body: formatComment({ article: articleResult.text, summary: summaryResult.text }),
      })
    } else {
      await octokit.rest.issues.createComment({
        owner,
        repo,
        issue_number: number,
        body: formatComment({ article: articleResult.text, summary: summaryResult.text }),
      })
    }
  } catch (error) {
    core.setFailed((error as any).message)
  }
}

main()
