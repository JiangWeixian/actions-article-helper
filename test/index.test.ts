import { it } from 'vitest'
import { createChatGPTAPI } from '../src/api'

it('chatgpt', async () => {
  // disable on github ci workflows
  if (process.env.CI) {
    return
  }
  const client = createChatGPTAPI(process.env.OPEN_API_KEY!)
  const result = await client.sendMessage('ping')
  console.log(result)
})
