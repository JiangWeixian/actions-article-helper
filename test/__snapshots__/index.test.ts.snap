// Vitest Snapshot v1

exports[`chatgpt 1`] = `"Replacement text: The feature of modern front-end framework that I most appreciate is the utilization of a filesystem-based routing system. This makes defining routes significantly simpler. For example, with Next.js, one can define routes as such:  \\"./pages/index.tsx\\" corresponds to \\"/\\", \\"./pages/issues/index.tsx\\" corresponds to \\"/issues\\" and \\"./pages/issues/id/[id].tsx\\" corresponds to \\"/issues/id/:id\\". With the introduction of Node.js 12, a new feature called \\"condition-exports\\" was brought to the table, allowing developers to specify package subexports more easily. For instance, when using \\"lodash,\\" one may import an API in such a manner: \\"import {isObject} from 'lodash',\\" and Webpack will bundle all \\"lodash\\" APIs into the final bundle, reducing size by utilizing \\"babel-plugin-import.\\" A new tool, \\"rollup-plugin-condition-exports@next,\\" allows library developers to take advantage of this same pattern syntax. By default, files located in the \\"src/exports\\" folder will be set up in the \\"exports\\" field in the \\"package.json.\\" In addition, Node.js supports \\"condition-names,\\" meaning developers can define different APIs for both Node and browser sides. This is useful when one wants to maintain the same API between both sides, but with different behaviors. Finally, \\"rollup-plugin-condition-exports@next\\" has its own condition name convention; files named \\"node,\\" \\"browser\\" or \\"deno\\" are set up in different conditions. Visit https://github.com/JiangWeixian/rollup-plugin-condition-exports for more information about this plugin."`;
